src/main/java/net/dries007/tfc/common/blocks/plant/KelpTreeBlock.java

```diff
@@ -0,0 +1,167 @@
+        }
+        else
+        {
+            updateFluid(worldIn, stateIn, currentPos);
+            boolean flag = facingState.is(TFCTags.Blocks.KELP_TREE) || (facing == Direction.DOWN && facingState.is(TFCTags.Blocks.SEA_BUSH_PLANTABLE_ON));
+            return stateIn.setValue(PROPERTY_BY_DIRECTION.get(facing), flag);
+        }
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public void tick(BlockState state, ServerWorld worldIn, BlockPos pos, Random rand)
+    {
+        if (!state.canSurvive(worldIn, pos))
+        {
+            worldIn.destroyBlock(pos, true);
+        }
+    }
+
+    @Override
+    public void playerWillDestroy(World worldIn, BlockPos pos, BlockState state, PlayerEntity player)
+    {
+        updateFluid(worldIn, state, pos);
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public boolean canSurvive(BlockState state, IWorldReader worldIn, BlockPos pos)
+    {
+        BlockState belowState = worldIn.getBlockState(pos.below());
+        //boolean attachedVertically = !KelpTreeFlowerBlock.isEmptyWaterBlock(worldIn, pos.above()) && !KelpTreeFlowerBlock.isEmptyWaterBlock(worldIn, pos.below());
+
+        for (Direction direction : Direction.Plane.HORIZONTAL)
+        {
+            BlockPos relativePos = pos.relative(direction);
+            if (worldIn.getBlockState(relativePos).getBlock().is(TFCTags.Blocks.KELP_BRANCH))
+            {
+                /*if (attachedVertically)
+                {
+                    return false;
+                }*/
+                //this might be causing overlapping structures to break each other. should re-evaluate
+                //if you study the structures ingame you'll see the algorithm precludes them from growing and then touching horizontally anyway
+                //so theoretically this should never be called anyway, I'm not sure
+
+                Block below = worldIn.getBlockState(relativePos.below()).getBlock();
+                if (below.is(TFCTags.Blocks.KELP_BRANCH) || below.is(TFCTags.Blocks.SEA_BUSH_PLANTABLE_ON))
+                {
+                    return true;
+                }
+            }
+        }
+        Block blockIn = belowState.getBlock();
+        return blockIn.is(TFCTags.Blocks.KELP_BRANCH) || blockIn.is(TFCTags.Blocks.SEA_BUSH_PLANTABLE_ON);
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public FluidState getFluidState(BlockState state)
+    {
+        return IFluidLoggable.super.getFluidState(state);
+    }
+
+    @Override
+    protected void createBlockStateDefinition(StateContainer.Builder<Block, BlockState> builder)
+    {
+        super.createBlockStateDefinition(builder);
+        builder.add(getFluidProperty());
+        builder.add(NORTH, EAST, SOUTH, WEST, UP, DOWN);
+    }
+
+    @Override
+    public FluidProperty getFluidProperty()
+    {
+        return getFluidPropertyAbstract();
+    }
+
+    public abstract FluidProperty getFluidPropertyAbstract();
```

Comment:
This doesn't make any sense, `getFluidProperty` already is abstract as it's provided by the interface implementation, and this method name is bad. Can just use `getFluidProperty` and remove all uses of `getFluidPropertyAbstract`